File: kpp-robotik/OLD_DO_NOT_SEE/kpp.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX_ENERGY = 1000;

struct Edge {
    string to;
    int w, o;
};

class Graph {
public:
    int n;
    unordered_map<string, vector<Edge>> adj;

    Graph(int n) : n(n) {}

    void add_edge(const string &u, const string &v, int w, int o) {
        adj[u].push_back({v, w, o});
        adj[v].push_back({u, w, o});
    }
};

struct State {
    int energy_used;
    int time;
    string node;
    int energy_left;
    vector<string> path;

    bool operator>(const State &other) const {
        return energy_used > other.energy_used;
    }
};

class Pathfinder {
private:
    Graph &graph;
    unordered_set<string> rest_points;
    unordered_set<string> charging_stations;
    int start_hour;
    string start, target;

    // visited[(node, energy_left, time)] = min_energy_used
    map<tuple<string,int,int>, int> visited;

public:
    Pathfinder(Graph &g,
               const string &s, const string &t,
               const unordered_set<string> &rest,
               const unordered_set<string> &charging,
               int hour)
        : graph(g), start(s), target(t),
          rest_points(rest), charging_stations(charging), start_hour(hour) {}

    void solve() {
        int start_time = start_hour * 60;

        priority_queue<State, vector<State>, greater<State>> pq;
        pq.push({0, start_time, start, MAX_ENERGY, {start}});

        while (!pq.empty()) {
            auto cur = pq.top(); pq.pop();

            int energy_used = cur.energy_used;
            int time = cur.time;
            string node = cur.node;
            int energy_left = cur.energy_left;
            vector<string> path = cur.path;

            // Goal check
            if (node == target) {
                print_result(energy_used, path);
                return;
            }

            auto state = make_tuple(node, energy_left, time);
            if (visited.count(state) && visited[state] <= energy_used) {
                continue;
            }
            visited[state] = energy_used;

            // Charging
            if (charging_stations.count(node)) {
                pq.push({energy_used, time, node, MAX_ENERGY, path});
            }

            // Rest
            if (rest_points.count(node)) {
                int current_hour = time / 60;
                int new_time = time;
                if (current_hour % 2 == 1) {
                    int minutes_to_next_hour = 60 - (time % 60);
                    new_time = time + minutes_to_next_hour;
                }
                pq.push({energy_used, new_time, node, energy_left, path});
            }

            // Explore neighbors
            for (auto &edge : graph.adj[node]) {
                int base_cost = edge.w + edge.o;
                int hour = time / 60;
                int cost;
                if (hour % 2 == 1) cost = int(base_cost * 1.3);
                else cost = int(base_cost * 0.8);

                if (cost <= energy_left) {
                    auto new_path = path;
                    new_path.push_back(edge.to);
                    pq.push({energy_used + cost, time + 2, edge.to,
                             energy_left - cost, new_path});
                }
            }
        }

        cout << "Robot gagal dalam mencapai tujuan :(\n";
    }

private:
    void print_result(int energy_used, const vector<string> &path) {
        cout << "Total energi minimum: " << energy_used << "\n";
        cout << "Jalur: ";
        for (size_t i = 0; i < path.size(); i++) {
            if (i) cout << " -> ";
            cout << path[i];
        }
        cout << "\nWaktu tiba:\n";
        for (size_t i = 0; i < path.size(); i++) {
            cout << path[i] << " (menit " << i*2 << ")\n";
        }
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;
    Graph graph(n);

    for (int i = 0; i < m; i++) {
        string u, v;
        int w, o;
        cin >> u >> v >> w >> o;
        graph.add_edge(u, v, w, o);
    }

    string S, T;
    cin >> S >> T;

    // Rest points
    unordered_set<string> rest_points;
    {
        string line;
        getline(cin, line); // buang newline sisa
        getline(cin, line);
        stringstream ss(line);
        string x;
        while (ss >> x) {
            if (x != "-") rest_points.insert(x);
        }
    }

    // Charging stations
    unordered_set<string> charging_stations;
    {
        string line;
        getline(cin, line);
        stringstream ss(line);
        string x;
        while (ss >> x) {
            if (x != "-") charging_stations.insert(x);
        }
    }

    string mechanic, electrical;
    getline(cin, mechanic);
    getline(cin, electrical);

    int start_hour;
    cin >> start_hour;

    Pathfinder solver(graph, S, T, rest_points, charging_stations, start_hour);
    solver.solve();

    return 0;
}

File: kpp-robotik/OLD_DO_NOT_SEE/kpp.py
# KPP Programming
# UKM Robotika ITS 2025
#
# Hendra Manudinata / 5027251051
# Teknologi Informasi / FTEIC

import heapq

# Energi total
# untuk setiap awal jalan dan recharge
MAX_ENERGY = 1000

class Graph:
    def __init__(self, n):
        self.n = n
        self.adj = { }  # adjacency list: node -> [(neighbor, w, o)]

    def add_edge(self, u, v, w, o):
        if u not in self.adj: self.adj[u] = []
        self.adj[u].append((v, w, o))
        if v not in self.adj: self.adj[v] = []
        self.adj[v].append((u, w, o))

def solve_it_dawg(
    graph,
    start,
    target,
    rest_points,
    charging_stations,
    start_hour
):
    time = start_hour * 60

    pq = []
    # Priority queue (pq)
    # (total_energy_used, time, node, energy_left, path)
    heapq.heappush(pq, (
        0, # total_energy_used
        time, # time
        start, # node
        MAX_ENERGY, # energy_left
        [start] # [path]
    ))

    # Simpan visited node
    visited = dict()

    # Loop terus selama PQ tidak kosong
    while pq:
        # Ambil antrian dari PQ
        energy_used, time, node, energy_left, path = heapq.heappop(pq)

        # Robot sudah sampai tujuan
        if node == target:
            print(f"Total energi minimum: {energy_used}")
            print("Jalur: " + " -> ".join(path))
            print("Waktu tiba:")
            for i, n in enumerate(path):
                print(f"{n} (menit {i*2})")  # asumsi tiap edge waktu = 2 menit di contoh
            return

        # Posisi saat ini
        state = (node, energy_left, time)

        # Jika node sudah dikunjungi
        # dan energi lebih hemat: Skip
        # Jika sebaliknya, simpan visited node saat ini
        if state in visited and visited[state] <= energy_used:
            continue
        else:
            visited[state] = energy_used

        # Charging = Isi ke maksimum
        if node in charging_stations:
            heapq.heappush(pq, (energy_used, time, node, MAX_ENERGY, path))

        # Rest point = Tunggu hingga jam genap
        if node in rest_points:
            current_hour = time // 60

            if current_hour % 2 == 1:  # kalau masih jam ganjil
                minutes_to_next_hour = 60 - (time % 60)
                new_time = time + minutes_to_next_hour
            else:
                # Jika sudah jam genap, tidak perlu rest
                # new_time = time + 1
                new_time = time

            heapq.heappush(pq, (energy_used, new_time, node, energy_left, path))

        # Explore neighbors
        # (Path finding)
        for v, w, o in graph.adj.get(node, []):
            base_cost = w + o

            # Efek waktu
            hour = time // 60
            if hour % 2 == 1:  # ganjil
                cost = int(base_cost * 1.3)
            else:  # genap
                cost = int(base_cost * 0.8)

            if cost <= energy_left:
                # Add path to PQ
                heapq.heappush(pq, (
                    energy_used + cost,
                    time + 2,
                    v,
                    energy_left - cost,
                    path + [v]
                ))

    # Jika PQ sudah kosong dan tidak sampai ke tujuan,
    # Robot gagal
    print("Robot gagal dalam mencapai tujuan :(")

def main():
    # Input: Graph Node & Edge
    n, m = map(int, input().split())
    graph = Graph(n)
    for _ in range(m):
        u, v, w, o = input().split()
        w, o = int(w), int(o)
        graph.add_edge(u, v, w, o)

    # S T endpoint
    S, T = input().split()

    # Rest points
    rest_points = set(input().split())
    if "-" in rest_points: rest_points = set()

    # Charging stations
    charging_stations = set(input().split())
    if "-" in charging_stations: charging_stations = set()

    # Mechanic & Electrical (UNUSED)
    mechanic = input()
    electrical = input()

    # Start minutes
    start_hour = int(input())

    solve_it_dawg(graph, S, T, rest_points, charging_stations, start_hour)

if __name__ == "__main__":
    main()
